### PR TITLE
Enum property

### PR DESCRIPTION
**Pull request by doctrino**
## Description
Please describe the change you have made.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: tests/tests_unit/test_data_classes/test_data_models/test_data_types.py [hunk 1]
```diff
@@ -37,12 +37,28 @@ class TestPropertyType:
...
+                        "description": "Time series with string data points.",
+                    },
+                    "numeric": {
+                        "name": "numeric",
+                        "description": "Time series with double floating point data points.",
+                    },
+                },
+            },
         ],
+        ids=lambda d: d.get("type", "unknown"),
```
Comment (doctrino) [inline tests/tests_unit/test_data_classes/test_data_models/test_data_types.py:54]: Just for nice info for IDE when running tests:
image

File: tests/tests_integration/test_api/test_data_modeling/test_graphql.py [hunk 2]
```diff
@@ -67,11 +67,11 @@ def test_apply_dm_raise_top_level_error(self, cognite_client: CogniteClient, int
...
         expected = """
-            type SomeType @view(version: "31e24bb338352b") {
+            type SomeType {
               someProp: String!
             }
 
-            type AnotherType @view(version: "855f080c9de22f") {
+            type AnotherType {
               anotherProp: String!
             }
```
Comment (doctrino) [inline tests/tests_integration/test_api/test_data_modeling/test_graphql.py:76]: Unrelated change. Seems to be a change in GraphQL.

File: CHANGELOG.md [hunk 3]
```diff
@@ -17,6 +17,14 @@ Changes are grouped as follows
 - `Fixed` for any bug fixes.
 - `Security` in case of vulnerabilities.
 
+## [7.54.3] - 2024-07-17
+### Fixed
+- Reintroduced `ListablePropertyType` that was removed in `7.37.0`. This is used to represent all container properties
+  that can be listed. This is an internal change and should not affect users of the SDK, but might affect developers
+  that have created libraries that depend on the SDK.
```
Comment (haakonvt) [inline CHANGELOG.md:24]: I think this can be removed from the changelog as it should be "internal SDK details". Only the actual types like `Date`, `DirectRelation` or `Float32` are explicitly made available to the user in e.g. `cognite/client/data_classes/data_modeling/__init__.py` and these still work the same.

File: cognite/client/data_classes/data_modeling/data_types.py [hunk 4]
```diff
@@ -73,7 +72,7 @@ def __load_unit_ref(data: dict) -> UnitReference | None:
     @classmethod
     def _load(cls, resource: dict[str, Any], cognite_client: CogniteClient | None = None) -> Self:
         type_ = resource["type"]
-        obj: Any
+        obj: Any  # This is a hack to avoid a mypy error
```
Comment (haakonvt) [inline cognite/client/data_classes/data_modeling/data_types.py]: I pushed a commit with an alternate solution (that you should feel free to reject 😉 )

File: cognite/client/data_classes/data_modeling/data_types.py [hunk 5]
```diff
@@ -117,32 +119,37 @@ def _load(cls, resource: dict[str, Any], cognite_client: CogniteClient | None =
...
+
+@dataclass
+class Text(ListablePropertyType):
     _type = "text"
     collation: str = "ucs_basic"
 
 
 @dataclass
-class Primitive(PropertyType, ABC): ...
+class Primitive(ListablePropertyType, ABC): ...
```
Comment (haakonvt) [inline cognite/client/data_classes/data_modeling/data_types.py:131]: Seems to be unused, can we remove it? Or does it belong in the type hierarchy?
```suggestion
class Primitive(ListablePropertyType, ABC): ...
```

File: cognite/client/data_classes/data_modeling/data_types.py [hunk 6]
```diff
@@ -106,6 +105,9 @@ def _load(cls, resource: dict[str, Any], cognite_client: CogniteClient | None =
                 # to avoid breaking changes. When used as a read object, the `list` will always be present.
                 is_list=resource.get("list", False),
             )
+        elif type_ == "enum":
+            values = {key: EnumValue.load(value) for key, value in resource["values"].items()}
```
Comment (haakonvt) [inline cognite/client/data_classes/data_modeling/data_types.py:108]: Can skip a stack frame & an isinstance check:
```suggestion
            values = {key: EnumValue._load(value) for key, value in resource["values"].items()}
```

File: tests/tests_unit/test_data_classes/test_data_models/test_data_types.py [hunk 7]
```diff
@@ -37,12 +37,28 @@ class TestPropertyType:
...
     )
     def test_load_dump(self, data: dict) -> None:
-        actual = PropertyType.load(data).dump(camel_case=True)
+        loaded = PropertyType.load(data)
+        dumped = loaded.dump(camel_case=True)
 
-        assert data == actual
+        assert isinstance(loaded, PropertyType)
+        assert data == dumped
```
Comment (haakonvt) [inline tests/tests_unit/test_data_classes/test_data_models/test_data_types.py:61]: Can this test be removed? I see `PropertyType` inherits from `CogniteObject` which I believe is covered by your automatic tests

File: cognite/client/data_classes/data_modeling/data_types.py (hunk 5 above)
Comment (doctrino) [reply cognite/client/data_classes/data_modeling/data_types.py:131]: We use it in the type hierarchy, and `pygen` and `neat` use it if memory serves.

File: cognite/client/data_classes/data_modeling/data_types.py (hunk 4 above)
Comment (doctrino) [reply cognite/client/data_classes/data_modeling/data_types.py]: Not pretty that either, but prettier

File: tests/tests_unit/test_data_classes/test_data_models/test_data_types.py (hunk 7 above)
Comment (doctrino) [reply tests/tests_unit/test_data_classes/test_data_models/test_data_types.py:61]: No, the issue is that you want to check that you load a `PropertyType` and not an `UnknownCogniteObject`. 

Note that when you only check `data == dumped` a `UnknownCogniteObject` will pass the type. Changed the check to make this clearer.

File: CHANGELOG.md [hunk 8]
```diff
@@ -17,6 +17,11 @@ Changes are grouped as follows
 - `Fixed` for any bug fixes.
 - `Security` in case of vulnerabilities.
 
+## [7.54.3] - 2024-07-17
+### Added
+- [Feature Preview] Support for `enum` as container property type in the data modeling APIs. Note that this is not
```
Comment (doctrino) [inline CHANGELOG.md:22]: Minor bump as it is an experimental feature.

File: tests/tests_unit/test_data_classes/test_data_models/test_data_types.py [hunk 9]
```diff
@@ -53,7 +69,7 @@ def test_load_ignore_unknown_properties(self) -> None:
     def test_load_dump_unkown_property(self) -> None:
         data = {"type": "unknowngibberish", "list": True, "unit": {"externalId": "known"}}
         obj = PropertyType.load(data)
-        assert isinstance(obj, UnknownPropertyType)
+        assert isinstance(obj, UnknownCogniteObject)
```
Comment (doctrino) [inline tests/tests_unit/test_data_classes/test_data_models/test_data_types.py:72]: Changed this one as `UnknownPropertyType` will likely be removed in next major version.

File: cognite/client/data_classes/data_modeling/data_types.py (hunk 5 above)
Comment (haakonvt) [reply cognite/client/data_classes/data_modeling/data_types.py:130]: Are you sure?
```
>>> from cognite.client.data_classes.data_modeling.data_types import Primitive
>>> from tests.utils import all_subclasses
>>> all_subclasses(Primitive)
```

File: tests/tests_unit/test_data_classes/test_data_models/test_data_types.py (hunk 7 above)
Comment (haakonvt) [reply tests/tests_unit/test_data_classes/test_data_models/test_data_types.py:61]: Right, makes sense! 👍

File: cognite/client/data_classes/data_modeling/data_types.py (hunk 5 above)
Comment (doctrino) [reply cognite/client/data_classes/data_modeling/data_types.py:130]: Is is used `isinstance(propy.type, Primitive)`

Comment (haakonvt) [reply cognite/client/data_classes/data_modeling/data_types.py:130]: Either the search in my editor suddenly stopped working or I can't find that line in the SDK.

**Review by haakonvt — APPROVED** [review]
LGTM!

File: CHANGELOG.md (hunk 8 above)
Comment (haakonvt) [reply CHANGELOG.md:22]: Minor?
